solr/contrib/s3-repository/src/java/org/apache/solr/s3/S3StorageClient.java

```diff
@@ -0,0 +1,568 @@
+        objectMetadata.setContentType(BLOB_DIR_CONTENT_TYPE);
+        objectMetadata.setContentLength(0);
+
+        // Create empty blob object with header
+        final InputStream im = ClosedInputStream.CLOSED_INPUT_STREAM;
+
+        try {
+            PutObjectRequest putRequest = new PutObjectRequest(bucketName, path, im, objectMetadata);
+            s3Client.putObject(putRequest);
+        } catch (AmazonClientException ase) {
+            throw handleAmazonException(ase);
+        }
+    }
+
+    /**
+     /**
+     * Delete files from S3 Blob Store. Deletion order is not guaranteed.
+     *
+     * @param paths Paths to files or blobs.
+     */
+    void delete(Collection<String> paths) throws S3Exception {
+        Set<String> entries = new HashSet<>();
+        for (String path : paths) {
+            entries.add(sanitizedFilePath(path));
+        }
+
+        deleteBlobs(entries);
+    }
+
+    /**
+     * Delete directory, all the files and sub-directories from S3.
+     *
+     * @param path Path to directory in S3.
+     */
+    void deleteDirectory(String path) throws S3Exception {
+        path = sanitizedDirPath(path);
+
+        Set<String> entries = new HashSet<>();
+        if (pathExists(path)) {
+            entries.add(path);
+        }
+
+        // Get all the files and subdirectories
+        entries.addAll(listAll(path));
+
+        deleteObjects(entries);
+    }
+
+    /**
+     * List all the files and sub-directories directly under given path.
+     *
+     * @param path Path to directory in S3.
+     * @return Files and sub-directories in path.
+     */
+    String[] listDir(String path) throws S3Exception {
+        path = sanitizedDirPath(path);
+
+        String prefix = path;
+        ListObjectsRequest listRequest = new ListObjectsRequest()
+            .withBucketName(bucketName)
+            .withPrefix(prefix)
+            .withDelimiter(BLOB_FILE_PATH_DELIMITER);
+
+        List<String> entries = new ArrayList<>();
+        try {
+            ObjectListing objectListing = s3Client.listObjects(listRequest);
+
+            while (true) {
+                List<String> files = objectListing.getObjectSummaries().stream()
+                        .map(S3ObjectSummary::getKey)
+                        .collect(Collectors.toList());
+                files.addAll(objectListing.getCommonPrefixes());
+                // This filtering is needed only for S3mock. Real S3 does not ignore the trailing '/' in the prefix.
+                files = files.stream()
+                    .filter(s -> s.startsWith(prefix))
+                    .map(s -> s.substring(prefix.length()))
+                    .filter(s -> !s.isEmpty())
+                    .filter(s -> {
+                        int slashIndex = s.indexOf(BLOB_FILE_PATH_DELIMITER);
+                        return slashIndex == -1 || slashIndex == s.length() - 1;
+                    })
+                    .map(s -> {
```

Comment:
What's the need for this? I think this removal of the trailing slash is what's breaking local restores for me. When I comment it out, it works again.

Reply:
I do this because `AbstractBackupRepositoryTest.testCanListFullOrEmptyDirectories()` excepts directories to not end in an ending slash.

I originally had changed the test to expect ending slashes, but it required a very large footprint through the rest of the backup-restore codebase that I thought it would be easier to just remove the ending slash here.

I'll do some tests though and see why this is breaking the s3 mock use case.

Reply:
Ok tested it out locally and it worked for me (backed up twice, listed the backups, deleted one of them, and restored the remaining one). This is all incremental, maybe the non-incremental breaks.... let me try that.

Would you mind sharing your error?

Reply:
I'm hitting `
solr/admin/collections?action=RESTORE&repository=s3&location=s3:&name=$BACKUP_NAME&collection=$RESTORED_COLLECTION_NAME`

The stacktrace isn't very helpful since it's thrown by `CollectionsHandler.invokeAction(CollectionsHandler.java:262)`. But the error that's thrown is from this check in https://github.com/apache/solr/blob/main/solr/core/src/java/org/apache/solr/core/backup/BackupManager.java#L129

Reply:
Ok I think I figured out that it's an error with non-incremental restores. It's not necessarily an issue with this method, but it is an issue that it's trying to see if a directory exists without adding the final slash to the directory name. Will make a fix that hopefully fixes all instances of this happening.

Reply:
Ahh replied without refreshing. Yes was able to reproduce that exactly. Will work on a fix 🙂

Reply:
Ok this should be fixed now.